File: src/SOP8/Base/Account/Base.Account.BLL/Process/Login/LoginManager.cs
using System;
using System.Collections.Generic;
using Base.Account.IBLL.Response;
using Base.Account.IBLL.Models;
using Base.Model.Account;
using dnsDapperDBUtil.DataAccessLayer.IDAL;
using Response.Resource;

namespace Base.Account.BLL.Process.Login
{
    using Resource;

    class LoginManager
    {
        // LoginFailLimit 이상 로그인 실패시(비밀번호 잘못 입력) 계정을 잠근다.(사용할수 없도록 한다.)
        private const int LoginFailLimit = 5;
        // 계정을 잠그는 시간
        private const int LockLoginMinutes = 30;

        private IDataManager m_dataManager = null;

        public LoginManager(IDataManager dataManager)
        {
            m_dataManager = dataManager;
        }

        public LoginResult Login(string strUserID, string strPW, string strSessionKey, bool autoLogin)
        {
            Grade grade = null;
            string strErrorMessage = null;

            LoginResult result = new LoginResult();

            // ID 값으로 유저를 검색
            string strCondition = string.Format("{0} = '{1}'", User.Fields.user_id, strUserID);
            User user = m_dataManager.GetSelect().SelectFirst<User>(strCondition, out strErrorMessage);

            if (user == null)
            {
                if (strErrorMessage != null)
                {
                    System.Diagnostics.Trace.WriteLine(strErrorMessage);

                    result.Success = false;
                    result.Message = ID.Get<ErrorMessage>("unknownUser").Value();
                    return result;
                }
                else
                {
                    result.Success = false;
                    result.Message = ID.Get<ErrorMessage>("invalidUserInfo").Value();
                    return result;
                }
            }

            if (CheckLoginFailCount(user, null, out strErrorMessage) == false)
            {
                result.Success = false;
                result.Message = strErrorMessage;
                return result;
            }

            if (user.password != strPW)
            {
                if (SetLoginFailCount(user, out strErrorMessage))
                {
                    if (strErrorMessage == null)
                        strErrorMessage = ID.Get<ErrorMessage>("invalidUserInfo").Value();
                }

                result.Success = false;
                result.Message = strErrorMessage;
                return result;
            }

            if (UpdateSession(m_dataManager, user.user_sn, strSessionKey, autoLogin, out strErrorMessage) == false)
            {
                result.Success = false;
                result.Message = strErrorMessage;
                return result;
            }

            strCondition = string.Format("{0} = {1}", Grade.Fields.grad_sn, user.grad_sn);
            grade = m_dataManager.GetSelect().SelectFirst<Grade>(strCondition, out strErrorMessage);

            if (grade == null)
            {
                if (strErrorMessage != null)
                    System.Diagnostics.Trace.WriteLine(strErrorMessage);

                result.Success = false;
                result.Message = ID.Get<ErrorMessage>("noneUserLevel").Value();
                return result;
            }

            // 로그인 성공 시 PasswordCode 초기화
            Dictionary<User.Fields, object> dicSets = new Dictionary<User.Fields, object>();
            dicSets[User.Fields.password_key] = null;

            strCondition = string.Format("{0} = {1}", User.Fields.user_sn, user.user_sn);

            if (m_dataManager.GetUpdate().Update<User, User.Fields>(dicSets, strCondition, out strErrorMessage) == false)
            {
                result.Success = false;
                result.Message = ID.Get<ErrorMessage>("failToInitLoginFail").Value();
                return result;
            }

            result.User = ApplicationUser.MakeUser(user, grade, strSessionKey);
            result.Success = true;
            return result;
        }

        public static bool UpdateSession(IDataManager dataManager, int nUserNo, string strSessionKey, bool autoLogin, out string strErrorMessage)
        {
            strErrorMessage = "";

            // 해당 계정 세션 유무 확인
            string strCondition = string.Format("{0} = {1}", Session.Fields.user_sn, nUserNo);
            Session session = dataManager.GetSelect().SelectFirst<Session>(strCondition, out strErrorMessage);

            if (session == null)
            {
                if (strErrorMessage != null)
                    return false;
            }
            else
            {
                // 있으면 삭제 후 생성, 없으면 생성
                strCondition = string.Format("{0} = {1}", Session.Fields.user_sn, nUserNo);

                if (dataManager.GetDelete().Delete<Session>(strCondition, out strErrorMessage) == false)
                    return false;
            }

            session = new Session();

            session.user_sn = nUserNo;
            session.session_key = strSessionKey;
            session.creat_de = DateTime.Now;
            session.updt_de = session.creat_de;
            session.atmc_login_yn = autoLogin;

            if (dataManager.GetCreate().Insert<Session>(session, out strErrorMessage) == false)
                return false;

            return true;
        }

        private bool CheckLoginFailCount(User user, string strTarget, out string strErrorMessage)
        {
            strErrorMessage = null;

            if (user.password_key == null)
                return true;

            int nIndex = user.password_key.IndexOf('_');

            if (nIndex <= 0)
                return true;

            if (strTarget == null)
                strTarget = "로그인";

            string strFailCount = user.password_key.Substring(0, nIndex).Trim();
            string strLastFailTime = user.password_key.Substring(nIndex + 1).Trim();

            int nFailCount;
            long lastFailTime;

            if (int.TryParse(strFailCount, out nFailCount) && long.TryParse(strLastFailTime, out lastFailTime))
            {
                // 로그인 실패횟수가 허용치를 초과한 경우
                if (nFailCount >= LoginFailLimit)
                {
                    DateTime lastTime = DateTime.FromBinary(lastFailTime);
                    TimeSpan span = DateTime.Now - lastTime;

                    if (span.TotalMinutes <= LockLoginMinutes)
                    {
                        int minutes = (int)(LockLoginMinutes - span.TotalMinutes + 0.9999);
                        strErrorMessage = string.Format(ID.Get<ErrorMessageFormat>("lockFail").Value(), LoginFailLimit, minutes, strTarget);
                        return false;
                    }
                    else
                    {
                        // LockLoginMinutes를 지났을 경우 초기화한다.
                        user.password_key = null;


                        Dictionary<User.Fields, object> dicSets = new Dictionary<User.Fields, object>();
                        dicSets[User.Fields.password_key] = null;

                        string strCondition = string.Format("{0} = {1}", User.Fields.user_sn, user.user_sn);
                        
                        if (m_dataManager.GetUpdate().Update<User, User.Fields>(dicSets, strCondition, out strErrorMessage) == false)
                        {
                            System.Diagnostics.Trace.WriteLine(strErrorMessage);
                            strErrorMessage = ID.Get<ErrorMessage>("failToInitLoginFail").Value();
                            return false;
                        }
                    }
                }
            }

            return true;
        }

        private bool SetLoginFailCount(User user, out string strErrorMessage)
        {
            strErrorMessage = null;
            bool changed = false;
            int nFailCount = 0;

            if (user.password_key != null)
            {
                int nIndex = user.password_key.IndexOf('_');

                if (nIndex > 0)
                {
                    string strFailCount = user.password_key.Substring(0, nIndex).Trim();

                    if (int.TryParse(strFailCount, out nFailCount))
                    {
                        nFailCount++;
                        DateTime dtNow = DateTime.Now;

                        user.password_key = string.Format("{0}_{1}", nFailCount, dtNow.ToBinary());
                        changed = true;

                        if (nFailCount >= LoginFailLimit)
                            strErrorMessage = string.Format(ID.Get<ErrorMessageFormat>("lockLoginFail").Value(), LoginFailLimit, LockLoginMinutes);
                    }
                }
            }

            if (changed == false)
            {
                nFailCount = 1;
                DateTime dtNow = DateTime.Now;
                user.password_key = string.Format("{0}_{1}", nFailCount, dtNow.ToBinary());
            }

            if (nFailCount < LoginFailLimit)
            {
                strErrorMessage = string.Format(ID.Get<ErrorMessageFormat>("lockLoginFail2").Value(), nFailCount, LoginFailLimit, LockLoginMinutes);
            }

            if (m_dataManager.GetUpdate().Update<User>(user, null, out strErrorMessage) == false)
            {
                System.Diagnostics.Trace.WriteLine(strErrorMessage);
                strErrorMessage = ID.Get<ErrorMessage>("failToUpdateLoginFail").Value();
            }

            return strErrorMessage == null;
        }
    }
}

File: src/SOP8/광양/EgovLogin/base-account-bll/src/main/java/egovframework/base/account/bll/process/login/LoginManager.java
package egovframework.base.account.bll.process.login;

import java.time.Duration;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;

import egovframework.base.account.ibll.response.LoginResult;
import egovframework.base.account.ibll.models.ApplicationUser;
import egovframework.base.model.account.Grade;
import egovframework.base.model.account.Session;
import egovframework.base.model.account.User;
import egovframework.base.account.bll.resource.ErrorMessage;
import egovframework.base.account.bll.resource.ErrorMessageFormat;
import egovframework.base.response.resource.ID;
import egovframework.base.generic.Holder;
import egovframework.base.dal.account.*;

public class LoginManager {

    private static final int LOGIN_FAIL_LIMIT = 5;
    private static final int LOCK_LOGIN_MINUTES = 30;

    private UserDAO userDao;
    private SessionDAO sessionDao;
    private GradeDAO gradeDao;

    public LoginManager(UserDAO userDao, SessionDAO sessionDao, GradeDAO gradeDao) {
        this.userDao = userDao;
        this.sessionDao = sessionDao;
        this.gradeDao = gradeDao;
    }

    public LoginResult login(String userID, String password, String sessionKey, boolean autoLogin) {
        String condition = String.format("%s = '%s'", User.Fields.user_id, userID);
        Holder<String> errorMessage = new Holder<String>();

        User user = this.userDao.selectUser(null, condition, errorMessage);
        //User user = dataManager.getSelect().selectFirst(User.class, condition);
        if (user == null) {
            if (errorMessage.value != null) {
                return new LoginResult(false, ID.get(ErrorMessage.class, "unknownUser").value());
            }
            else {
                return new LoginResult(false, ID.get(ErrorMessage.class, "invalidUserInfo").value());
            }
        }

        if (!checkLoginFailCount(user, "로그인", errorMessage)) {
            return new LoginResult(false, errorMessage.value);
        }

        if (!user.getPassword().equals(password)) {
            if (setLoginFailCount(user, errorMessage)) {
                if (errorMessage.value == null)
                    errorMessage.value = ID.get(ErrorMessage.class, "invalidUserInfo").value();
            }

            return new LoginResult(false, errorMessage.value);
        }

        if (!updateSession(this.sessionDao, user.getUser_sn(), sessionKey, autoLogin, errorMessage)) {
            return new LoginResult(false, errorMessage.value);
        }

        condition = String.format("%s = %d", Grade.Fields.grad_sn, user.getGrad_sn());
        Grade grade = this.gradeDao.selectGrade(null, condition, errorMessage);
        //Grade grade = dataManager.getSelect().selectFirst(Grade.class, condition);

        if (grade == null) {
            return new LoginResult(false, ID.get(ErrorMessage.class, "noneUserLevel").value());
        }

        // Reset password_key
        Map<User.Fields, Object> updates = new HashMap<>();
        updates.put(User.Fields.password_key, null);

        condition = String.format("%s = %d", User.Fields.user_sn, user.getUser_sn());

        if (!this.userDao.updateUser(updates, condition, errorMessage)) {
        //if (!dataManager.getUpdate().update(User.class, updates, condition)) {
            return new LoginResult(false, ID.get(ErrorMessage.class, "failToInitLoginFail").value());
        }

        LoginResult result = new LoginResult(true, "");
        result.setUser(ApplicationUser.makeUser(user, grade, sessionKey));
        return result;
    }

    public static boolean updateSession(SessionDAO sessionDao, int userNo, String sessionKey, boolean autoLogin, Holder<String> errorMessage) {
        // 해당 계정 세션 유무 확인
        String condition = String.format("%s = %d", Session.Fields.user_sn, userNo);
        Session session = sessionDao.selectSession(null, condition, errorMessage);
        //Session session = dataManager.getSelect().selectFirst(Session.class, condition);

        if (session != null && !sessionDao.deleteSession(condition, errorMessage)) {
            return false;
        }

        session = new Session();
        session.setUser_sn(userNo);
        session.setSession_key(sessionKey);
        session.setCreat_de(LocalDateTime.now());
        session.setUpdt_de(session.getCreat_de());
        session.setAtmc_login_yn(autoLogin);

        return sessionDao.insertSession(session, errorMessage);
    }

    private boolean checkLoginFailCount(User user, String target, Holder<String> errorMessage) {
        errorMessage.value = null;

        if (user.getPassword_key() == null || !user.getPassword_key().contains("_")) {
            return true;
        }

        if (target == null)
            target = "로그인";

        String[] parts = user.getPassword_key().split("_");
        int failCount = Integer.parseInt(parts[0]);
        long failTimeBinary = Long.parseLong(parts[1]);
        LocalDateTime failTime = LocalDateTime.ofEpochSecond(failTimeBinary, 0, java.time.ZoneOffset.UTC);

        if (failCount >= LOGIN_FAIL_LIMIT) {
            long minutesElapsed = Duration.between(failTime, LocalDateTime.now()).toMinutes();
            if (minutesElapsed <= LOCK_LOGIN_MINUTES) {
                int remaining = (int) (LOCK_LOGIN_MINUTES - minutesElapsed + 0.9999);
                errorMessage.value = String.format(ID.get(ErrorMessageFormat.class, "lockFail").value(),
                        LOGIN_FAIL_LIMIT, remaining, target);
                return false;
            } else {
                user.setPassword_key(null);
                Map<User.Fields, Object> updates = new HashMap<>();
                updates.put(User.Fields.password_key, null);
                String condition = String.format("%s = %d", User.Fields.user_sn, user.getUser_sn());

                if (!this.userDao.updateUser(updates, condition, errorMessage)) {
                    errorMessage.value = ID.get(ErrorMessage.class, "failToInitLoginFail").value();
                    return false;
                }
            }
        }

        return true;
    }

    private boolean setLoginFailCount(User user, Holder<String> errorMessage) {
        errorMessage.value = null;

        int failCount = 1;
        boolean updated = false;
        LocalDateTime now = LocalDateTime.now();

        if (user.getPassword_key() != null && user.getPassword_key().contains("_")) {
            String[] parts = user.getPassword_key().split("_");
            failCount = Integer.parseInt(parts[0]) + 1;
            updated = true;
        }

        user.setPassword_key(String.format("%d_%d", failCount, now.toEpochSecond(java.time.ZoneOffset.UTC)));

        if (failCount < LOGIN_FAIL_LIMIT) {
            errorMessage.value = String.format(ID.get(ErrorMessageFormat.class, "lockLoginFail").value(), LOGIN_FAIL_LIMIT, LOCK_LOGIN_MINUTES);
            /*dataManager.setLastError(String.format(ID.get(ErrorMessageFormat.class, "lockLoginFail").value(),
                    LOGIN_FAIL_LIMIT, LOCK_LOGIN_MINUTES));*/
        } else {
            errorMessage.value = String.format(ID.get(ErrorMessageFormat.class, "lockLoginFail2").value(),
                    failCount, LOGIN_FAIL_LIMIT, LOCK_LOGIN_MINUTES);
        }

        if (this.userDao.updateUser(user, errorMessage) == false)
        {
            errorMessage.value = ID.get(ErrorMessageFormat.class, "failToUpdateLoginFail").value();
            return false;
        }

        return true;
    }
}